docs/caching.md

```diff
@@ -0,0 +1,281 @@
+# Caching
+
+Streamlit provides a caching mechanism that allows your app to stay performant even when loading data from the web, manipulating large datasets, or performing expensive computations. This is done with the [`@st.cache`](api.html#streamlit.cache) decorator.
+
+When you mark a function with the [`@st.cache`](api.html#streamlit.cache) decorator, it tells Streamlit that whenever the function is called it needs to check a few things:
+
+1. The input parameters that you called the function with
+2. The value of any external variable used in the function
+3. The body of the function
+4. The body of any function used inside the cached function
+
+If this is the first time Streamlit has seen these items with these exact values and in this exact combination and order, it runs the function and stores the result in a local cache. Think of the cache as a simple in-memory key-value store, where the key is a hash of all of the above and the value is the actual output object passed by reference.
+
+Then, next time the cached function is called, if the key hasn’t changed, Streamlit will just skip executing the function altogether and, instead, return the output previously stored in the cache.
+
+The way Streamlit keeps track of changes in these components is through hashing.
+
+Finally, [`@st.cache`](api.html#streamlit.cache) supports arguments to configure the cache's behavior. You can find more information on those in our [API reference](api.md).
+
+Let's take a look at a few examples that illustrate how caching works in a Streamlit app.
+
+## Example 1: Basic usage
+
+For starters, let's take a look at a sample app that has a function that performs an expensive, long-running computation. Without caching, this function is rerun each time the app is refreshed, leading to a poor user experience. Copy this code into a new app and try it out yourself:
+
+```python
+import streamlit as st
+import time
+
+def expensive_computation(a, b):
+    time.sleep(2)  # 👈 This makes the function take 2s to run
+    return a * b
+
+a = 2
+b = 21
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+```
+
+You'll notice that this app is laggy. If you press **R** to rerun the app, it gets worse, since the calculation is rerun from start to finish with each refresh. This isn't a great experience.
```

Comment:
This paragraph isn't very clear either. How about:
> Try pressing **R** to rerun the app, and notice how long it takes for the result to show up. This is because `expensive_computation(a, b)` is being re-executed every time the app runs. This isn't a great experience.

Reply:
Changed to your text.